homeassistant/components/owfs.py

```diff
@@ -0,0 +1,468 @@
+                cls = classes[dev.family]
+            except KeyError:
+                cls = UnknownDevice
+            obj = cls(hass, dev)
+            await obj._init(True)
+            objs.append(obj)
+        if objs:
+            async_add_entities(objs)
+
+
+class OWFSDevice(Entity):
+    """Representation of (an attribute of an) OWFS device"""
+    _attr = None
+    _val = None
+    _do_poll = 0
+    _next_poll = 0
+    _events = None  # list: remembered events
+
+    def __init__(self, hass: HomeAssistant, dev: "trio_owfs.device.Device",
+                 config: dict = {}):
+        self.hass = hass
+        self.dev = dev
+        self._events = None
+
+        name = config.get(CONF_NAME)
+        if name is None:
+            name = "{} {}".format(self.__class__.__name__, dev.id)
+        self._name = name
+        self._config = config
+
+        # save a ref to this device
+        try:
+            reg = dev.__objects
+        except AttributeError:
+            dev.__objects = reg = WeakValueDictionary()
+        attr = config.get(CONF_ATTRIBUTE)
+        if attr is None:
+            attr = "_default"
+        else:
+            self._attr = attr
+        reg[attr] = self
+
+    async def _init(self, without_config: bool):
+        """Async part of device initialization.
+        This code processes queued events and ensures that no more events
+        may be queued."""
+        events, self._events = self._events, None
+        for e in events:
+            await self.process_event(e)
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def should_poll(self):
+        """overrides this Entity property.
+        OWFS doesn't require polling by HASS.
+        """
+        if not self._do_poll:
+            return False
+        t = time()
+        if self._next_poll > t:
+            return False
+        self._next_poll = t + self._do_poll
+        return True
+
+    @property
+    def unique_id(self):
+        """overrides this Entity property.
+        Return the OWFS device ID (considered unique).
+        """
+        res = self.dev.id
+        if self._attr is not None:
+            res += "_" + self._attr.replace('/', '_')
+        return res
+
+    @property
+    def available(self):
+        """overrides this Entity property.
+        Returns True if the device's bus address is known.
+        """
+        return self.dev.bus is not None
+
+    def __repr__(self):
+        rd = repr(self.dev)[1:]
+        if self.name and not self.name.endswith(self.dev.id):
+            rd = self.name + ' ' + rd
+        return "<%s:%s>" % (self.__class__.__name__, rd)
+
+    async def process_event(self, evt: "trio_owfs.event.DeviceEvent"):
+        """handle an event from OWFS.
+
+        If the event is an alarm, trigger a message.
+        """
+        from trio_owfs.event import DeviceAlarm, DeviceValue
+
+        if isinstance(evt, DeviceAlarm):
+            self.hass.bus.async_fire(
```

Comment:
Entities shouldn't fire events. Instead, alarms should be handled by the component, which can fire events when necessary.